crates/bevy_ui/src/flex/mod.rs

```diff
@@ -37,257 +40,376 @@ impl Default for FlexSurface {
+        }
     }
 
-    // compute layouts
-    flex_surface.compute_window_layouts();
+    // Find all entities that are in this stacking context
+    let mut stacking_context = Vec::new();
+    fill_stacking_context(entity, &mut stacking_context, transform_z_query);
 
-    let physical_to_logical_factor = 1. / logical_to_physical_factor;
+    stacking_context.sort_by_key(|(_entity, z_index)| *z_index);
 
-    let to_logical = |v| (physical_to_logical_factor * v as f64) as f32;
+    // Set their z transform evenly spaced in 0..N
+    for (i, (entity, _z_index)) in stacking_context.iter().enumerate() {
+        let mut transform = transform_z_query
+            .get_component_mut::<Transform>(*entity)
+            .unwrap();
+        transform.translation.z = (i + 1) as f32;
+    }
+    // Set the scale of the origin so that other entities are effectively spaced in 0..1 in the origin's local transform
+    let mut transform = transform_z_query
+        .get_component_mut::<Transform>(entity)
+        .unwrap();
+    transform.scale.z = 1.0 / (stacking_context.len() + 1) as f32;
+}
 
-    for (entity, mut node, mut transform, parent, parent_flags, transform_flags) in
-        node_transform_query.iter_mut()
-    {
-        let layout = flex_surface.get_layout(entity).unwrap();
-        node.size = Vec2::new(
-            to_logical(layout.size.width),
-            to_logical(layout.size.height),
-        );
-        let position = &mut transform.translation;
-        position.x = to_logical(layout.location.x + layout.size.width / 2.0);
-        position.y = to_logical(layout.location.y + layout.size.height / 2.0);
-        if parent_flags.changed() || transform_flags.changed() {
-            if let Some(parent) = parent {
-                if let Ok(parent_layout) = flex_surface.get_layout(parent.0) {
-                    position.x -= to_logical(parent_layout.size.width / 2.0);
-                    position.y -= to_logical(parent_layout.size.height / 2.0);
+fn fill_stacking_context(
+    entity: Entity,
+    stacking_context: &mut Vec<(Entity, i16)>,
+    transform_z_query: &Query<(&Style, Option<&Children>, Option<&Parent>, &mut Transform)>,
+) {
+    if let Ok(children) = transform_z_query.get_component::<Children>(entity) {
+        for &child in children.iter() {
+            let z_index = &transform_z_query
+                .get_component::<Style>(child)
+                .unwrap()
+                .z_index;
+            match z_index {
+                ZIndex::None => stacking_context.push((child, 0)),
+                ZIndex::Auto => {
+                    stacking_context.push((child, 0));
+                    fill_stacking_context(child, stacking_context, transform_z_query);
                 }
-            }
+                ZIndex::Some(i) => stacking_context.push((child, *i)),
+            };
         }
     }
 }
+
+pub fn garbage_collection_system(
+    mut flex_surface: ResMut<FlexSurface>,
+    style_query: Query<&Style>,
```

Comment:
Why not use `style_query.removed::<Style>()` here instead of iterating over everything?

Reply:
I thought that `style_query.removed::<Style>()` only queries for entities that had their `Style` component removed but still exist, I'm glad to be wrong!